src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -53,6 +57,10 @@ def make_install_req_from_link(link, parent):
             hashes=parent.hash_options
         ),
     )
+    if ireq.link is None:
+        ireq.link = link
+    # TODO: Handle wheel cache resolution.
+    return ireq
```

Comment:
@uranusjr here is a track to explore. Is it correct that what you want to do here is create a new InstallRequirement that is identical to `parent` except for `link`? In that case it is important that the new ireq preserves parent's `original_link`.

Reply:
We did originally have code that cloned the `InstallRequirement` by creating a new object and copying all(>?) of the state across. It was ugly, but I believe it also had some problems. @uranusjr can you remember what went wrong with that approach? I'm wondering if we should see whether switching back to that approach works any better than what we currently do...

Reply:
The old approach makes the PEP 610 implementation think the cloned `InstallRequirement` is a direct one, because we would generate the `Candidate`’s `InstallRequirement` with a directly URL, i.e. candidates requested by `a==1.0.0` and `a @ file://.../a-1.0.0-...whl` are undistinguishable in the previous approach.

Reply:
Ah yes, that was the problem. I had a suspicion it would be something simple. I need a cup of tea to kick-start my brain :-)